subprojects/language-jvm/src/main/java/org/gradle/api/tasks/compile/ProviderAwareForkOptions.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.compile;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import org.gradle.api.Incubating;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.Optional;
+import org.gradle.process.CommandLineArgumentProvider;
+import org.gradle.util.internal.CollectionUtils;
+
+import java.util.List;
+
+/**
+ * Fork options for compilation that can accept user-defined {@link CommandLineArgumentProvider} objects.
+ *
+ * Only take effect if {@code fork} is {@code true}.
+ *
+ * @since 7.1
+ */
+@Incubating
+public class ProviderAwareForkOptions extends BaseForkOptions {
```

Comment:
There is also the interface `JavaForkOptions`, which also have `getJvmArgumentProviders()`. Can this type easily implement `JavaForkOptions`?

Reply:
The main reason I didn't do this is because `JavaForkOptions` exposes working directory and environment as mutable values.  Environment is maybe a debatable case, but I think we definitely don't want the user changing the working directory of the compiler daemons.  We would likely have to cut up that interface some to handle this.  There are also similar, but different methods on these objects which would need to be duplicated/deprecated if we extend off of `JavaForkOptions` (i.e. `JavaForkOptions.getMaxHeapSize()` vs `ForkOptions.getMemoryMaximumSize()`).

Having said that, I think this is something we should do (or at least attempt).  I don't see these issues as blockers so much as things that raise the complexity of the change and probably warrant tackling it as a separate effort.

Reply:
Yeah, I vaguely remember being in a similar place when we introduced the command line argument providers in the first place. Let's tackle this as a separate effort.